internal/auth/ldap/repository_auth_method_update.go

```diff
@@ -0,0 +1,563 @@
+type factoryFunc func(ctx context.Context, publicId string, idx int, i any) (any, error)
+
+// supportedFactories are the currently supported factoryFunc for value objects
+var supportedFactories = map[voName]factoryFunc{
+	CertificateVO: func(ctx context.Context, publicId string, idx int, i any) (any, error) {
+		str := fmt.Sprintf("%s", i)
+		return NewCertificate(ctx, publicId, str)
+	},
+	UrlVO: func(ctx context.Context, publicId string, idx int, i any) (any, error) {
+		u, err := url.Parse(fmt.Sprintf("%s", i))
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, "ldap.urlFactory")
+		}
+		return NewUrl(ctx, publicId, idx+1, u)
+	},
+}
+
+// valueObjectChanges takes the new and old list of VOs (value objects) and
+// using the dbMasks/nullFields it will return lists of VOs where need to be
+// added and deleted in order to reconcile auth method's value objects.
+func valueObjectChanges(
+	ctx context.Context,
+	publicId string,
+	valueObjectName voName,
+	newVOs,
+	oldVOs,
+	dbMask,
+	nullFields []string,
+) (add []any, del []any, e error) {
+	const op = "valueObjectChanges"
+	switch {
+	case publicId == "":
+		return nil, nil, errors.New(ctx, errors.InvalidParameter, op, "missing public id")
+	case !validVoName(valueObjectName):
+		return nil, nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("invalid value object name: %s", valueObjectName))
+	case !strutil.StrListContains(dbMask, string(valueObjectName)) && !strutil.StrListContains(nullFields, string(valueObjectName)):
+		return nil, nil, nil
+	case len(strutil.RemoveDuplicates(newVOs, false)) != len(newVOs):
+		return nil, nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("duplicate new %s", valueObjectName))
+	case len(strutil.RemoveDuplicates(oldVOs, false)) != len(oldVOs):
+		return nil, nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("duplicate old %s", valueObjectName))
+	}
+
+	factory, ok := supportedFactories[valueObjectName]
+	if !ok {
+		return nil, nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("unsupported factory for value object: %s", valueObjectName))
+	}
+
+	foundVOs := map[string]int{}
+	for i, a := range oldVOs {
+		foundVOs[a] = i
+	}
+	var adds []any
+	var deletes []any
+	if strutil.StrListContains(nullFields, string(valueObjectName)) {
+		deletes = make([]any, 0, len(oldVOs))
+		for i, v := range oldVOs {
+			deleteObj, err := factory(ctx, publicId, i, v)
+			if err != nil {
+				return nil, nil, errors.Wrap(ctx, err, op)
+			}
+			deletes = append(deletes, deleteObj)
+			delete(foundVOs, v)
+		}
+	}
+	if strutil.StrListContains(dbMask, string(valueObjectName)) {
+		adds = make([]any, 0, len(newVOs))
+		for i, v := range newVOs {
+			if _, ok := foundVOs[v]; ok {
+				delete(foundVOs, v)
+				continue
+			}
+			obj, err := factory(ctx, publicId, i, v)
+			if err != nil {
+				return nil, nil, errors.Wrap(ctx, err, op)
+			}
+			adds = append(adds, obj)
+			delete(foundVOs, v)
+		}
+	}
+	if len(foundVOs) > 0 {
```

Comment:
I think you can remove this since the for won't range over anything for an empty map

Reply:
ty!

Reply:
I'm not sure that it's empty.  If the VO isn't in the nullFields and it's in the dbMask, but doesn't match an existing oldVO then it will still be in the foundVOs map.